Build fails with Intel 16 and Intel 17 compilers on Linux

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => 3.4.1 (and master branch)
- Operating System / Platform => Linux CentOS 7.4 / x86_64
- Compiler => Intel 16.0.2, 17.0.1

##### Detailed description

Hi,

OpenCV doesn't build when compiled with some Intel compiler versions. I have tried the following three:

16.0.2 - make fails
17.0.1 - make fails
18.0.2 - make succeeds

The build fails at different points, depending on which version of cmake I'm using at the time.

Example failure message (Intel 17.0.1, cmake 2.8.12.2):
```
[ 38%] Building CXX object modules/imgproc/CMakeFiles/opencv_imgproc.dir/src/smooth.cpp.o
cd /nobackup/foo/o/opencv-3.4.1/build/modules/imgproc && /apps/developers/compilers/intel/17.0.1/1/default/compilers_and_libraries_2017.1.132/linux/bin/intel64/icpc   -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -fsigned-char -fp-model precise  -msse3 -O3 -DNDEBUG  -fPIC -I/nobackup/foo/o/opencv-3.4.1/build/3rdparty/ippicv/ippicv_lnx/include -I/nobackup/foo/o/opencv-3.4.1/build/3rdparty/ippicv/ippiw_lnx/include -I/nobackup/foo/o/opencv-3.4.1/build -I/nobackup/foo/o/opencv-3.4.1/modules/imgproc/include -I/nobackup/foo/o/opencv-3.4.1/modules/imgproc/src -I/nobackup/foo/o/opencv-3.4.1/build/modules/imgproc -I/nobackup/foo/o/opencv-3.4.1/modules/core/include    -o CMakeFiles/opencv_imgproc.dir/src/smooth.cpp.o -c /nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp
/nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp(2676): error: more than one conversion function from "<unnamed>::ufixedpoint32" to "cv::uint8_t={uint8_t={unsigned char}}" applies:
            function "<unnamed>::ufixedpoint32::operator double() const"
            function "<unnamed>::ufixedpoint32::operator float() const"
          dst[i] = m[0] * src0[i];
                   ^

/nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp(2693): error: more than one conversion function from "const <unnamed>::ufixedpoint16" to "cv::uint8_t={uint8_t={unsigned char}}" applies:
            function "<unnamed>::ufixedpoint16::operator double() const"
            function "<unnamed>::ufixedpoint16::operator float() const"
          dst[i] = src0[i];
                   ^

/nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp(2740): error: more than one conversion function from "<unnamed>::ufixedpoint32" to "cv::uint8_t={uint8_t={unsigned char}}" applies:
            function "<unnamed>::ufixedpoint32::operator double() const"
            function "<unnamed>::ufixedpoint32::operator float() const"
          dst[i] = m[0] * src[0][i] + m[1] * src[1][i] + m[2] * src[2][i];
                   ^

/nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp(2816): error: more than one conversion function from "<unnamed>::ufixedpoint32" to "cv::uint8_t={uint8_t={unsigned char}}" applies:
            function "<unnamed>::ufixedpoint32::operator double() const"
            function "<unnamed>::ufixedpoint32::operator float() const"
          dst[i] = m[0] * src[0][i] + m[1] * src[1][i] + m[2] * src[2][i] + m[3] * src[3][i] + m[4] * src[4][i];
                   ^

/nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp(2919): error: more than one conversion function from "<unnamed>::ufixedpoint32" to "cv::uint8_t={uint8_t={unsigned char}}" applies:
            function "<unnamed>::ufixedpoint32::operator double() const"
            function "<unnamed>::ufixedpoint32::operator float() const"
          dst[i] = val;
                   ^

compilation aborted for /nobackup/foo/o/opencv-3.4.1/modules/imgproc/src/smooth.cpp (code 2)
make[2]: *** [modules/imgproc/CMakeFiles/opencv_imgproc.dir/src/smooth.cpp.o] Error 2
make[2]: Leaving directory `/nobackup/foo/o/opencv-3.4.1/build'
make[1]: *** [modules/imgproc/CMakeFiles/opencv_imgproc.dir/all] Error 2
make[1]: Leaving directory `/nobackup/foo/o/opencv-3.4.1/build'
make: *** [all] Error 2
```

Example failure message (Intel 17.0.1, cmake 3.10.0):
```
[ 36%] Building CXX object modules/imgproc/CMakeFiles/opencv_imgproc.dir/src/resize.cpp.o
/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(568): error: more than one conversion function from "<unnamed>::ufixedpoint16" to "uint8_t={unsigned char}" applies:
            function "<unnamed>::ufixedpoint16::operator double() const"
            function "<unnamed>::ufixedpoint16::operator float() const"
          *(dst++) = *(src++);
                     ^

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(617): error: more than one conversion function from "<unnamed>::ufixedpoint32" to "uint8_t={unsigned char}" applies:
            function "<unnamed>::ufixedpoint32::operator double() const"
            function "<unnamed>::ufixedpoint32::operator float() const"
          *(dst++) = (uint8_t)(*(src++) * m[0] + *(src1++) * m[1]);
                              ^

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(550): error: more than one conversion function from "<unnamed>::fixedpoint32" to "schar={signed char}" applies:
            function "<unnamed>::fixedpoint32::operator double() const"
            function "<unnamed>::fixedpoint32::operator float() const"
          dst[i] = src[i];
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineSet(FT *, ET *, int) [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32]" at line 693
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with
                      ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=schar={signed char}, interpolation=<unnamed>::interpolationLinear<schar={signed char}>]" at line 3798

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(579): error: more than one conversion function from "<unnamed>::fixedpoint32::WT" to "schar={signed char}" applies:
            function "<unnamed>::fixedpoint64::operator double() const"
            function "<unnamed>::fixedpoint64::operator float() const"
          dst[i] = res;
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineResize<ET,FT,n>(FT *, size_t={unsigned long}, FT *, ET *, int) [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, n=2]" at line 710
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with
                      ET=schar={signed char}, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=schar={signed char}, interpolation=<unnamed>::interpolationLinear<schar={signed char}>]" at line 3798

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(550): error: more than one conversion function from "<unnamed>::ufixedpoint32" to "ushort={unsigned short}" applies:
            function "<unnamed>::ufixedpoint32::operator double() const"
            function "<unnamed>::ufixedpoint32::operator float() const"
          dst[i] = src[i];
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineSet(FT *, ET *, int) [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32]" at line 693
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with
                      ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=ushort={unsigned short}, interpolation=<unnamed>::interpolationLinear<ushort={unsigned short}>]" at line 3799

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(579): error: more than one conversion function from "<unnamed>::ufixedpoint32::WT" to "ushort={unsigned short}" applies:
            function "<unnamed>::ufixedpoint64::operator double() const"
            function "<unnamed>::ufixedpoint64::operator float() const"
          dst[i] = res;
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineResize<ET,FT,n>(FT *, size_t={unsigned long}, FT *, ET *, int) [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, n=2]" at line 710
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with
                      ET=ushort={unsigned short}, FT=<unnamed>::ufixedpoint32, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=ushort={unsigned short}, interpolation=<unnamed>::interpolationLinear<ushort={unsigned short}>]" at line 3799

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(550): error: more than one conversion function from "<unnamed>::fixedpoint32" to "short" applies:
            function "<unnamed>::fixedpoint32::operator double() const"
            function "<unnamed>::fixedpoint32::operator float() const"
          dst[i] = src[i];
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineSet(FT *, ET *, int) [with ET=short, FT=<unnamed>::fixedpoint32]" at line 693
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=short, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=short, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=short, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with ET=short,
                      FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=short, interpolation=<unnamed>::interpolationLinear<short>]" at line 3800

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(579): error: more than one conversion function from "<unnamed>::fixedpoint32::WT" to "short" applies:
            function "<unnamed>::fixedpoint64::operator double() const"
            function "<unnamed>::fixedpoint64::operator float() const"
          dst[i] = res;
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineResize<ET,FT,n>(FT *, size_t={unsigned long}, FT *, ET *, int) [with ET=short, FT=<unnamed>::fixedpoint32, n=2]" at line 710
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=short, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=short, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=short, FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with ET=short,
                      FT=<unnamed>::fixedpoint32, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=short, interpolation=<unnamed>::interpolationLinear<short>]" at line 3800

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(550): error: more than one conversion function from "<unnamed>::fixedpoint64" to "int" applies:
            function "<unnamed>::fixedpoint64::operator double() const"
            function "<unnamed>::fixedpoint64::operator float() const"
          dst[i] = src[i];
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineSet(FT *, ET *, int) [with ET=int, FT=<unnamed>::fixedpoint64]" at line 693
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=int, FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=int, FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=int, FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with ET=int,
                      FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=int, interpolation=<unnamed>::interpolationLinear<int>]" at line 3801

/scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp(579): error: more than one conversion function from "<unnamed>::fixedpoint64::WT" to "int" applies:
            function "<unnamed>::fixedpoint64::operator double() const"
            function "<unnamed>::fixedpoint64::operator float() const"
          dst[i] = res;
                   ^
          detected during:
            instantiation of "void <unnamed>::vlineResize<ET,FT,n>(FT *, size_t={unsigned long}, FT *, ET *, int) [with ET=int, FT=<unnamed>::fixedpoint64, n=2]" at line 710
            instantiation of "void <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::operator()(const cv::Range &) const [with ET=int, FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 676
            implicit generation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::~resize_bitExactInvoker() [with ET=int, FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 676
            instantiation of class "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len> [with ET=int, FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 676
            instantiation of "<unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::resize_bitExactInvoker(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, int *, int *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::fixedpoint *, int, int, int, int, <unnamed>::resize_bitExactInvoker<ET, FT, interp_y_len>::hResizeFunc) [with ET=int,
                      FT=<unnamed>::fixedpoint64, interp_y_len=2]" at line 774
            instantiation of "void <unnamed>::resize_bitExact<ET,interpolation>(const uchar={unsigned char} *, size_t={unsigned long}, int, int, uchar={unsigned char} *, size_t={unsigned long}, int, int, int, double, double) [with ET=int, interpolation=<unnamed>::interpolationLinear<int>]" at line 3801

compilation aborted for /scratch/foo/pp/opencv-3.4.1/modules/imgproc/src/resize.cpp (code 2)
make[2]: *** [modules/imgproc/CMakeFiles/opencv_imgproc.dir/src/resize.cpp.o] Error 2
make[1]: *** [modules/imgproc/CMakeFiles/opencv_imgproc.dir/all] Error 2
make: *** [all] Error 2
```

<!-- your description -->

##### Steps to reproduce

Make Intel 17.0.1 or 17.0.2 available to shell environment. Then:
```
tar xf opencv-3.4.1.tar.gz
cd opencv-3.4.1

mkdir build
cd build

export VERBOSE=1
export CC=icc 
export CXX=icpc 
export FC=ifort 
export F77=ifort 
export F90=ifort 

cmake ..
make
```
<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

--- a/modules/imgproc/src/fixedpoint.inl.hpp
+++ b/modules/imgproc/src/fixedpoint.inl.hpp
@@ -91,9 +91,14 @@ public:
     CV_ALWAYS_INLINE fixedpoint64 operator << (int n) const { return fixedpoint64(val << n); }
     CV_ALWAYS_INLINE bool operator == (const fixedpoint64& val2) const { return val == val2.val; }
     template <typename ET>
-    CV_ALWAYS_INLINE operator ET() const { return cv::saturate_cast<ET>((int64_t)fixedround((uint64_t)val) >> fixedShift); }
+    CV_ALWAYS_INLINE ET saturate_cast() const { return cv::saturate_cast<ET>((int64_t)fixedround((uint64_t)val) >> fixedShift); }
     CV_ALWAYS_INLINE operator double() const { return (double)val / (1LL << fixedShift); }
     CV_ALWAYS_INLINE operator float() const { return (float)val / (1LL << fixedShift); }
+    CV_ALWAYS_INLINE operator uint8_t() const { return saturate_cast<uint8_t>(); }
+    CV_ALWAYS_INLINE operator int8_t() const { return saturate_cast<int8_t>(); }
+    CV_ALWAYS_INLINE operator uint16_t() const { return saturate_cast<uint16_t>(); }
+    CV_ALWAYS_INLINE operator int16_t() const { return saturate_cast<int16_t>(); }
+    CV_ALWAYS_INLINE operator int32_t() const { return saturate_cast<int32_t>(); }
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE fixedpoint64 zero() { return fixedpoint64(); }
     static CV_ALWAYS_INLINE fixedpoint64 one() { return fixedpoint64((int64_t)(1LL << fixedShift)); }
@@ -151,9 +156,14 @@ public:
     CV_ALWAYS_INLINE ufixedpoint64 operator << (int n) const { return ufixedpoint64(val << n); }
     CV_ALWAYS_INLINE bool operator == (const ufixedpoint64& val2) const { return val == val2.val; }
     template <typename ET>
-    CV_ALWAYS_INLINE operator ET() const { return cv::saturate_cast<ET>(fixedround(val) >> fixedShift); }
+    CV_ALWAYS_INLINE ET saturate_cast() const { return cv::saturate_cast<ET>(fixedround(val) >> fixedShift); }
     CV_ALWAYS_INLINE operator double() const { return (double)val / (1LL << fixedShift); }
     CV_ALWAYS_INLINE operator float() const { return (float)val / (1LL << fixedShift); }
+    CV_ALWAYS_INLINE operator uint8_t() const { return saturate_cast<uint8_t>(); }
+    CV_ALWAYS_INLINE operator int8_t() const { return saturate_cast<int8_t>(); }
+    CV_ALWAYS_INLINE operator uint16_t() const { return saturate_cast<uint16_t>(); }
+    CV_ALWAYS_INLINE operator int16_t() const { return saturate_cast<int16_t>(); }
+    CV_ALWAYS_INLINE operator int32_t() const { return saturate_cast<int32_t>(); }
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE ufixedpoint64 zero() { return ufixedpoint64(); }
     static CV_ALWAYS_INLINE ufixedpoint64 one() { return ufixedpoint64((uint64_t)(1ULL << fixedShift)); }
@@ -198,10 +208,15 @@ public:
     CV_ALWAYS_INLINE fixedpoint32 operator << (int n) const { return fixedpoint32(val << n); }
     CV_ALWAYS_INLINE bool operator == (const fixedpoint32& val2) const { return val == val2.val; }
     template <typename ET>
-    CV_ALWAYS_INLINE operator ET() const { return cv::saturate_cast<ET>((int32_t)fixedround((uint32_t)val) >> fixedShift); }
+    CV_ALWAYS_INLINE ET saturate_cast() const { return cv::saturate_cast<ET>((int32_t)fixedround((uint32_t)val) >> fixedShift); }
     CV_ALWAYS_INLINE operator fixedpoint64() const { return (int64_t)val << (fixedpoint64::fixedShift - fixedShift); }
     CV_ALWAYS_INLINE operator double() const { return (double)val / (1 << fixedShift); }
     CV_ALWAYS_INLINE operator float() const { return (float)val / (1 << fixedShift); }
+    CV_ALWAYS_INLINE operator uint8_t() const { return saturate_cast<uint8_t>(); }
+    CV_ALWAYS_INLINE operator int8_t() const { return saturate_cast<int8_t>(); }
+    CV_ALWAYS_INLINE operator uint16_t() const { return saturate_cast<uint16_t>(); }
+    CV_ALWAYS_INLINE operator int16_t() const { return saturate_cast<int16_t>(); }
+    CV_ALWAYS_INLINE operator int32_t() const { return saturate_cast<int32_t>(); }
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE fixedpoint32 zero() { return fixedpoint32(); }
     static CV_ALWAYS_INLINE fixedpoint32 one() { return fixedpoint32((1 << fixedShift)); }
@@ -242,10 +257,15 @@ public:
     CV_ALWAYS_INLINE ufixedpoint32 operator << (int n) const { return ufixedpoint32(val << n); }
     CV_ALWAYS_INLINE bool operator == (const ufixedpoint32& val2) const { return val == val2.val; }
     template <typename ET>
-    CV_ALWAYS_INLINE operator ET() const { return cv::saturate_cast<ET>(fixedround(val) >> fixedShift); }
+    CV_ALWAYS_INLINE ET saturate_cast() const { return cv::saturate_cast<ET>(fixedround(val) >> fixedShift); }
     CV_ALWAYS_INLINE operator ufixedpoint64() const { return (uint64_t)val << (ufixedpoint64::fixedShift - fixedShift); }
     CV_ALWAYS_INLINE operator double() const { return (double)val / (1 << fixedShift); }
     CV_ALWAYS_INLINE operator float() const { return (float)val / (1 << fixedShift); }
+    CV_ALWAYS_INLINE operator uint8_t() const { return saturate_cast<uint8_t>(); }
+    CV_ALWAYS_INLINE operator int8_t() const { return saturate_cast<int8_t>(); }
+    CV_ALWAYS_INLINE operator uint16_t() const { return saturate_cast<uint16_t>(); }
+    CV_ALWAYS_INLINE operator int16_t() const { return saturate_cast<int16_t>(); }
+    CV_ALWAYS_INLINE operator int32_t() const { return saturate_cast<int32_t>(); }
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE ufixedpoint32 zero() { return ufixedpoint32(); }
     static CV_ALWAYS_INLINE ufixedpoint32 one() { return ufixedpoint32((1U << fixedShift)); }
@@ -284,10 +304,15 @@ public:
     CV_ALWAYS_INLINE fixedpoint16 operator << (int n) const { return fixedpoint16((int16_t)(val << n)); }
     CV_ALWAYS_INLINE bool operator == (const fixedpoint16& val2) const { return val == val2.val; }
     template <typename ET>
-    CV_ALWAYS_INLINE operator ET() const { return cv::saturate_cast<ET>((int16_t)fixedround((uint16_t)val) >> fixedShift); }
+    CV_ALWAYS_INLINE ET saturate_cast() const { return cv::saturate_cast<ET>((int16_t)fixedround((uint16_t)val) >> fixedShift); }
     CV_ALWAYS_INLINE operator fixedpoint32() const { return (int32_t)val << (fixedpoint32::fixedShift - fixedShift); }
     CV_ALWAYS_INLINE operator double() const { return (double)val / (1 << fixedShift); }
     CV_ALWAYS_INLINE operator float() const { return (float)val / (1 << fixedShift); }
+    CV_ALWAYS_INLINE operator uint8_t() const { return saturate_cast<uint8_t>(); }
+    CV_ALWAYS_INLINE operator int8_t() const { return saturate_cast<int8_t>(); }
+    CV_ALWAYS_INLINE operator uint16_t() const { return saturate_cast<uint16_t>(); }
+    CV_ALWAYS_INLINE operator int16_t() const { return saturate_cast<int16_t>(); }
+    CV_ALWAYS_INLINE operator int32_t() const { return saturate_cast<int32_t>(); }
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE fixedpoint16 zero() { return fixedpoint16(); }
     static CV_ALWAYS_INLINE fixedpoint16 one() { return fixedpoint16((int16_t)(1 << fixedShift)); }
@@ -324,10 +349,15 @@ public:
     CV_ALWAYS_INLINE ufixedpoint16 operator << (int n) const { return ufixedpoint16((uint16_t)(val << n)); }
     CV_ALWAYS_INLINE bool operator == (const ufixedpoint16& val2) const { return val == val2.val; }
     template <typename ET>
-    CV_ALWAYS_INLINE operator ET() const { return cv::saturate_cast<ET>(fixedround(val) >> fixedShift); }
+    CV_ALWAYS_INLINE ET saturate_cast() const { return cv::saturate_cast<ET>(fixedround(val) >> fixedShift); }
     CV_ALWAYS_INLINE operator ufixedpoint32() const { return (uint32_t)val << (ufixedpoint32::fixedShift - fixedShift); }
     CV_ALWAYS_INLINE operator double() const { return (double)val / (1 << fixedShift); }
     CV_ALWAYS_INLINE operator float() const { return (float)val / (1 << fixedShift); }
+    CV_ALWAYS_INLINE operator uint8_t() const { return saturate_cast<uint8_t>(); }
+    CV_ALWAYS_INLINE operator int8_t() const { return saturate_cast<int8_t>(); }
+    CV_ALWAYS_INLINE operator uint16_t() const { return saturate_cast<uint16_t>(); }
+    CV_ALWAYS_INLINE operator int16_t() const { return saturate_cast<int16_t>(); }
+    CV_ALWAYS_INLINE operator int32_t() const { return saturate_cast<int32_t>(); }
     CV_ALWAYS_INLINE bool isZero() { return val == 0; }
     static CV_ALWAYS_INLINE ufixedpoint16 zero() { return ufixedpoint16(); }
     static CV_ALWAYS_INLINE ufixedpoint16 one() { return ufixedpoint16((uint16_t)(1 << fixedShift)); }

--- a/modules/imgproc/src/fixedpoint.inl.hpp
+++ b/modules/imgproc/src/fixedpoint.inl.hpp
@@ -55,9 +55,11 @@ public:
     CV_ALWAYS_INLINE fixedpoint64 operator * (const int32_t& val2) const { return operator *(fixedpoint64(val2)); }
     CV_ALWAYS_INLINE fixedpoint64 operator * (const fixedpoint64& val2) const
     {
-        uint64_t uval = (uint64_t)((val ^ (val >> 63)) - (val >> 63));
-        uint64_t umul = (uint64_t)((val2.val ^ (val2.val >> 63)) - (val2.val >> 63));
-        int64_t ressign = (val >> 63) ^ (val2.val >> 63);
+        bool sign_val = val < 0;
+        bool sign_mul = val2.val < 0;
+        uint64_t uval = sign_val ? (uint64_t)(-val) : (uint64_t)val;
+        uint64_t umul = sign_mul ? (uint64_t)(-val2.val) : (uint64_t)val2.val;
+        bool ressign = sign_val ^ sign_mul;
 
         uint64_t sh0   = fixedround((uval & 0xFFFFFFFF) * (umul & 0xFFFFFFFF));
         uint64_t sh1_0 = (uval >> 32)        * (umul & 0xFFFFFFFF);
@@ -67,25 +69,24 @@ public:
         uint64_t val0_h = (sh2   & 0xFFFFFFFF) + (sh1_0 >> 32) + (sh1_1 >> 32) + (val0_l >> 32);
         val0_l &= 0xFFFFFFFF;
 
-        if ( (sh2 >> 32) || (val0_h >> ressign ? 32 : 31) )
-            return (ressign ? ~(int64_t)0x7FFFFFFFFFFFFFFF : (int64_t)0x7FFFFFFFFFFFFFFF);
+        if (sh2 > CV_BIG_INT(0x7FFFFFFF) || val0_h > CV_BIG_INT(0x7FFFFFFF))
+            return (int64_t)(ressign ? CV_BIG_UINT(0x8000000000000000) : CV_BIG_INT(0x7FFFFFFFFFFFFFFF));
 
         if (ressign)
         {
-            val0_l = (~val0_l + 1) & 0xFFFFFFFF;
-            val0_h = val0_l ? ~val0_h : (~val0_h + 1);
+            return -(int64_t)(val0_h << 32 | val0_l);
         }
         return (int64_t)(val0_h << 32 | val0_l);
     }
     CV_ALWAYS_INLINE fixedpoint64 operator + (const fixedpoint64& val2) const
     {
         int64_t res = val + val2.val;
-        return ((val ^ res) & (val2.val ^ res)) >> 63 ? ~(res & ~0x7FFFFFFFFFFFFFFF) : res;
+        return (int64_t)(((val ^ res) & (val2.val ^ res)) < 0 ? ~(res & CV_BIG_UINT(0x8000000000000000)) : res);
     }
     CV_ALWAYS_INLINE fixedpoint64 operator - (const fixedpoint64& val2) const
     {
         int64_t res = val - val2.val;
-        return ((val ^ val2.val) & (val ^ res)) >> 63 ? ~(res & ~0x7FFFFFFFFFFFFFFF) : res;
+        return (int64_t)(((val ^ val2.val) & (val ^ res)) < 0 ? ~(res & CV_BIG_UINT(0x8000000000000000)) : res);
     }
     CV_ALWAYS_INLINE fixedpoint64 operator >> (int n) const { return fixedpoint64(val >> n); }
     CV_ALWAYS_INLINE fixedpoint64 operator << (int n) const { return fixedpoint64(val << n); }
@@ -138,15 +139,15 @@ public:
         uint64_t val0_h = (sh2 & 0xFFFFFFFF) + (sh1_0 >> 32) + (sh1_1 >> 32) + (val0_l >> 32);
         val0_l &= 0xFFFFFFFF;
 
-        if ((sh2 >> 32) || (val0_h >> 32))
-            return ((uint64_t)0xFFFFFFFFFFFFFFFF);
+        if (sh2 > CV_BIG_INT(0xFFFFFFFF) || val0_h > CV_BIG_INT(0xFFFFFFFF))
+            return (uint64_t)CV_BIG_UINT(0xFFFFFFFFFFFFFFFF);
 
-        return val0_h << 32 | val0_l;
+        return (val0_h << 32 | val0_l);
     }
     CV_ALWAYS_INLINE ufixedpoint64 operator + (const ufixedpoint64& val2) const
     {
         uint64_t res = val + val2.val;
-        return (val > res) ? (uint64_t)0xFFFFFFFFFFFFFFFF : res;
+        return (uint64_t)((val > res) ? CV_BIG_UINT(0xFFFFFFFFFFFFFFFF) : res);
     }
     CV_ALWAYS_INLINE ufixedpoint64 operator - (const ufixedpoint64& val2) const
     {
@@ -197,12 +198,12 @@ public:
     CV_ALWAYS_INLINE fixedpoint32 operator + (const fixedpoint32& val2) const
     {
         int32_t res = val + val2.val;
-        return ((val ^ res) & (val2.val ^ res)) >> 31 ? ~(res & ~0x7FFFFFFF) : res;
+        return (int64_t)((val ^ res) & (val2.val ^ res)) >> 31 ? ~(res & ~0x7FFFFFFF) : res;
     }
     CV_ALWAYS_INLINE fixedpoint32 operator - (const fixedpoint32& val2) const
     {
         int32_t res = val - val2.val;
-        return ((val ^ val2.val) & (val ^ res)) >> 31 ? ~(res & ~0x7FFFFFFF) : res;
+        return (int64_t)((val ^ val2.val) & (val ^ res)) >> 31 ? ~(res & ~0x7FFFFFFF) : res;
     }
     CV_ALWAYS_INLINE fixedpoint32 operator >> (int n) const { return fixedpoint32(val >> n); }
     CV_ALWAYS_INLINE fixedpoint32 operator << (int n) const { return fixedpoint32(val << n); }

